### PR TITLE
feat: checkout status check before starting application

### PR DESCRIPTION
**Pull request by axel7083**
### What does this PR do?

Empower the user with information on the repository status when starting an application

Scenarios

(1) repository is not cloned
- cloning repository

(2) repository is cloned and behind
- git fetch
- show warning to the user, do you want to update ?
- 'Update' => git pull
- 'Continue' => use as it is
- 'Cancel' => throw error

(3) repository is cloned and detached
- git fetch
- show warning to the user, do you want to reset ?
- 'Reset' => delete folder & git clone & git checkout
- 'Continue' => use as it is
- 'Cancel' => throw error

### Screenshot / video of UI

https://github.com/projectatomic/ai-studio/assets/42176370/415a46df-823f-4cdb-a0c6-580cde86d696

### What issues does this PR fix or reference?

Fixes https://github.com/projectatomic/ai-studio/issues/243

### How to test this PR?

- [x] unit tests has been provided

### CODE REVIEW

File: packages/backend/src/managers/applicationManager.ts [hunk 1]
```diff
@@ -118,10 +118,12 @@ export class ApplicationManager extends Publisher<ApplicationState[]> {
         ref: recipe.ref,
         targetDirectory: localFolder,
       };
+      console.log('[applicationManager] before doCheckout');
```
Comment (benoitf) [inline packages/backend/src/managers/applicationManager.ts:121]: will be removed I suppose ?

File: packages/backend/src/managers/applicationManager.ts [hunk 2]
```diff
@@ -118,10 +118,12 @@ export class ApplicationManager extends Publisher<ApplicationState[]> {
...
       await this.doCheckout(gitCloneInfo, {
         'recipe-id': recipe.id,
         'model-id': model.id,
       });
+      console.log('[applicationManager] after doCheckout');
```
Comment (benoitf) [inline packages/backend/src/managers/applicationManager.ts:126]: same here ?

File: packages/backend/src/managers/gitManager.ts [hunk 3]
```diff
@@ -33,4 +35,137 @@ export class GitManager {
...
+        gitCloneInfo.targetDirectory,
+        gitCloneInfo.repository,
+        gitCloneInfo.ref,
+      );
+
+      if (result.ok) {
+        return;
+      }
+
+      const error = `The repository "${gitCloneInfo.repository}" seems to already be cloned and is not matching the expected configuration: ${result.error}`;
```
Comment (benoitf) [inline packages/backend/src/managers/gitManager.ts:75]:
```suggestion
      const error = `The repository "${gitCloneInfo.repository}" appears to have already been cloned and does not match the expected configuration: ${result.error}`;
```

File: packages/backend/src/managers/gitManager.ts [hunk 4]
```diff
@@ -33,4 +35,137 @@ export class GitManager {
...
+
+      // Ask user
+      const selected = await window.showWarningMessage(
+        `${error} By continuing, the AI application may not run as expected. `,
+        'Cancel',
+        'Continue',
+        result.updatable ? 'Update' : 'Reset',
+      );
+
+      if (selected === undefined || selected === 'Cancel') {
```
Comment (benoitf) [inline packages/backend/src/managers/gitManager.ts:85]:
```suggestion
      if (!selected || selected === 'Cancel') {
```

File: packages/backend/src/managers/gitManager.ts [hunk 5]
```diff
@@ -33,4 +35,137 @@ export class GitManager {
...
+
+      if (selected === 'Update') {
+        // Update
+        await this.pull(gitCloneInfo.targetDirectory);
+        return;
+      }
+
+      if (selected === 'Reset') {
+        rmSync(gitCloneInfo.targetDirectory, { recursive: true });
+      }
```
Comment (benoitf) [inline packages/backend/src/managers/gitManager.ts:101]: looks like we could avoid the return instructions using `if else if else if` pattern

File: packages/backend/src/managers/gitManager.ts [hunk 6]
```diff
@@ -33,4 +35,137 @@ export class GitManager {
...
+    }
+
+    if (status.deleted.length > 0) {
+      return { error: 'The local repository has created files.' };
+    }
+
+    if (status.ahead !== 0) {
+      return {
+        error: `The local repository has ${status.ahead} commit(s) ahead.`,
+      };
```
Comment (benoitf) [inline packages/backend/src/managers/gitManager.ts:151]: else if

File: packages/backend/src/managers/gitManager.ts [hunk 7]
```diff
@@ -33,4 +35,137 @@ export class GitManager {
...
+
+    // Ensure working with a clean branch
+    if (!status.isClean()) {
+      return { error: 'The local repository is not clean.' };
+    }
+
+    // If we are not in HEAD
+    if (status.behind !== 0) {
+      return { ok: true, updatable: true };
+    }
```
Comment (benoitf) [inline packages/backend/src/managers/gitManager.ts:169]: else if ?

Comment (axel7083) [reply packages/backend/src/managers/gitManager.ts]: yes

File: packages/backend/src/managers/gitManager.ts (hunk 6 above)
Comment (axel7083) [reply packages/backend/src/managers/gitManager.ts]: yes

File: packages/backend/src/managers/applicationManager.ts (hunk 2 above)
Comment (axel7083) [reply packages/backend/src/managers/applicationManager.ts]: 👍

File: packages/backend/src/managers/applicationManager.ts (hunk 1 above)
Comment (axel7083) [reply packages/backend/src/managers/applicationManager.ts]: 👍

**Review by benoitf — APPROVED** [review]

**Review by jeffmaury — COMMENTED** [review]
I have the dialog even if I did not change the git repo neither the catalog so to me it should be updated

**Comment by axel7083** [comment]
> I have the dialog even if I did not change the git repo neither the catalog so to me it should be updated

I will open an issue and investigate, I don't seems to have it on my machine

**Comment by vrothberg** [comment]
I am facing the dialog at the moment and have troubles understanding what it means.  The message that the "local repository is detached" seems like an implementation detail to me that I do not understand.  I am afraid that users would be overwhelmed with this message and would always click on "Reset".

I don't have a good proposal for an alternative but wanted to highlight the challenge when being faced with the prompt.

Screenshot 2024-03-21 at 11 04 21

**Comment by axel7083** [comment]
> I am facing the dialog at the moment and have troubles understanding what it means. The message that the "local repository is detached" seems like an implementation detail to me that I do not understand. I am afraid that users would be overwhelmed with this message and would always click on "Reset".
> 
> I don't have a good proposal for an alternative but wanted to highlight the challenge when being faced with the prompt.
> 
> Screenshot 2024-03-21 at 11 04 21

See for fix https://github.com/projectatomic/ai-studio/pull/612

**Pull request merged**